Support larger page size in NVMM

mashoodk opened the pull request.

mashoodk committed:
Add more prints if shelf base directory creation failed

mashoodk committed:
Merge remote-tracking branch 'upstream/devel' into devel

gautham-bhat-k commented on src/shelf_mgmt/shelf_file.cc:
```diff
@@ -124,6 +128,7 @@ ErrorCode ShelfFile::Destroy() {
 ErrorCode ShelfFile::Truncate(off_t length) {
     TRACE();
     int ret;
+    length = ((length + device_pagesize_ - 1)/device_pagesize_)*device_pagesize_;
```
Can't we use **round_up_with_zero_check** function here ?

gautham-bhat-k commented on src/memory_manager.cc:
```diff
@@ -92,9 +93,12 @@ int StartNVMM(std::string base, std::string user) {
           return SHELF_BASE_DIR_CREATE_FAILED;
         }
       } catch (boost::filesystem::filesystem_error const &err) {
+	  std::cout << "NVMM: failed to create base dir at " << config.ShelfBase 
```
Is cout required here ?

mashoodk replied:
modified it to cerr, it is not most efficient method. But something is better than nothing. It gives a reason why starting of memory server is failing in OpenFAM.

BA-Ramya commented on src/common/epoch_shelf.cc:
```diff
@@ -241,8 +245,8 @@ ErrorCode EpochShelf::Close()
 
     //LOG(fatal) << "unregister fam atomic " << (uint64_t)addr_;
     fam_atomic_unregister_region(addr_, kShelfSize);
```
fam_atomic_unregister_region should also use length instead of kShelfSize. Isn't it?

BA-Ramya commented on src/common/root_shelf.cc:
```diff
@@ -144,7 +146,7 @@ ErrorCode RootShelf::Create()
     //LOG(fatal) << "unregister fam atomic " << (uint64_t)addr;
     fam_atomic_unregister_region(addr, kShelfSize);
```
Similar to comment in epoch_shelf.cc. Should we use length instead of kShelfSize?

BA-Ramya commented on src/memory_manager.cc:
```diff
@@ -121,6 +122,14 @@ int StartNVMM(std::string base, std::string user) {
 
+  //Set config.PageSize here
+  struct statvfs vfs;
+  if (statvfs(config.ShelfBase.c_str(), &vfs)==0) {
```
statvfs could fail for different error codes like EACCES, EBADF, ENAMETOOLONG, etc.
Do we need to address that instead of just setting getpagesize() as default?

mashoodk replied:
For now I have added error logging. Since usage of different page size for is rare for now, we can move ahead without acting upon.

BA-Ramya reviewed (approved):
Changes look fine to me

mashoodk committed:
Support larger page size in NVMM

Currently NVMM supports only system page size, it is possible
for mount points to have larger page sizes. Enable larger
pagesizes.

Signed-off-by: Mashood Abdulla <mashood.abdulla@hpe.com>
